lib/api_watch_diff.ml

```diff
@@ -81,69 +96,123 @@ let diff_value ~typing_env ~val_name ~reference ~current =
   | _, _ -> Some ()
   | exception Includecore.Dont_match _ -> Some ()
 
-let compare_values ~reference ~current =
+let rec diff_items curr_mod_name ~reference ~current =
   let env = env_setup ~ref_sig:reference ~curr_sig:current in
-  let ref_values = extract_values reference in
-  let curr_values = extract_values current in
-  FieldMap.merge
-    (fun val_name ref_opt curr_opt ->
-      match (ref_opt, curr_opt) with
-      | None, None -> None
-      | None, Some curr_vd -> Some (Value (val_name, Added curr_vd))
-      | Some ref_vd, None -> Some (Value (val_name, Removed ref_vd))
-      | Some ref_vd, Some curr_vd -> (
-          let value_differs =
-            diff_value ~typing_env:env ~val_name ~reference:ref_vd
-              ~current:curr_vd
-          in
-          match value_differs with
-          | None -> None
-          | Some _ ->
-              Some
-                (Value (val_name, Modified { ref = ref_vd; current = curr_vd }))
-          ))
-    ref_values curr_values
-  |> FieldMap.bindings |> List.map snd
-
-let diff_interface ~reference ~current =
-  let value_diffs = compare_values ~reference ~current in
-  if value_diffs = [] then
-    let typing_env = Env.empty in
-    let coercion1 () =
-      Includemod.signatures typing_env ~mark:Mark_both reference current
-    in
-    let coercion2 () =
-      Includemod.signatures typing_env ~mark:Mark_both current reference
-    in
-    match (coercion1 (), coercion2 ()) with
-    | Tcoerce_none, Tcoerce_none -> []
-    | _, _ -> [ Any ]
-    | exception Includemod.Error _ -> [ Any ]
-  else value_diffs
-
-let to_text_diff (diff_result : diff list) : Diffutils.Diff.t =
-  let vd_to_string name vd =
-    let buf = Buffer.create 256 in
-    let formatter = Format.formatter_of_buffer buf in
-    Printtyp.value_description (Ident.create_local name) formatter vd;
-    Format.pp_print_flush formatter ();
-    Buffer.contents buf
+  let ref_items = extract_items reference in
+  let curr_items = extract_items current in
+  let item_changes =
+    Sig_item_map.merge
+      (fun (item_type, name) ref_opt curr_opt ->
+        match (item_type, ref_opt, curr_opt) with
```

Comment:
There are a lot of things to simplify here.

I noticed that `diff_value` still returns a `unit option`. That's a legacy from a time before we had a representation for diffs. We should make it return a `value_description change option` at the very least. The point of making it return a `unit option` initially was so that we could replace the `unit` in there by a diff type for value descriptions.

You can make this pattern-matching much simpler to read by flattening it, there's not need for nested pattern matching here. For example the value case can be turned into something like:
```ocaml
| Value_item, None, None -> None
| Value_item, Some ref, None -> Some (Value {name; change = Removed ref})
| Value item, None, Some curr -> Some (Value {name; change = Added ref})
| Value_item, Some ref, Some curr ->
  let change = diff_value ~typing_env ... in
  Option.map (fun change -> Value {name; change}) change
 | Module_item, None, None, -> None
  ...
```

Alternatively, you could extract a `diff_value_item` function with the following signature:
```ocaml
val diff_value_item : name: string -> ref: value_description option -> curr: value_description option -> item_change option
```
so that the diff_items function just has to dispatch based on item_type and call other functions to keep things simple.